Source/zip/man/zip.1
.\" =========================================================================
.\" Copyright (c) 1990-2009 Info-ZIP.  All rights reserved.
.\"
.\" See the accompanying file LICENSE, version 2009-Jan-2 or later
.\" (the contents of which are also included in zip.h) for terms of use.
.\" If, for some reason, all these files are missing, the Info-ZIP license
.\" also may be found at:  ftp://ftp.info-zip.org/pub/infozip/license.html
.\" ==========================================================================
.\"
.\" zip.1 by Mark Adler, Jean-loup Gailly and  R. P. C. Rodgers
.\" updated by E. Gordon for Zip 3.0 (8 May 2005, 24 December 2006,
.\" 4 February 2007, 27 May 2007, 4 June 2007 by EG; 12 June 2007 by CS;
.\" 30 August 2007, 27 April 2008, 25 May 2008, 27 May 2008 by EG,
.\" 7 June 2008 by SMS and EG; 12 June 2008 by EG; 7 September 2008 by EG;
.\" 3 April 2009 by EG; 11 August 2009 by EG)
.\"
.TH ZIP 1L "12 August 2009 (v3.1)" Info-ZIP
.SH NAME
zip \- package and compress (archive) files
.SH SYNOPSIS
.B zip
.RB [\- aABcdDeEfFghjklLmoqrRSTuvVwXyz!@$ ]
[\-\-longoption ...]
.RB [\- b " path]"
.RB [\- n " suffixes]"
.RB [\- t " date]"
.RB [\- tt " date]"
[\fIzipfile\fR [\fIfile\fR \.\|.\|.]]
[\fB-xi\fR list]
.PP
.B zipcloak
(see separate man page)
.PP
.B zipnote
(see separate man page)
.PP
.B zipsplit
(see separate man page)
.PP
Note:  Command line processing in
.I zip
has been changed to support long options and handle all
options and arguments more consistently.  Some old command
lines that depend on command line inconsistencies may no longer
work.
.SH DESCRIPTION
.I zip
is a compression and file packaging utility for Unix, VMS, MSDOS,
OS/2, Windows 9x/NT/XP, Minix, Atari, Macintosh, MVS, z/OS, Amiga,
Acorn RISC, and other OS.  It is analogous to a combination of
the Unix commands
.IR tar (1)
and
.IR compress (1)
(or \fItar\fP(1) and \fIgzip\fP(1))
and is compatible with PKZIP (Phil Katz's ZIP for MSDOS systems)
and other major zip utilities.
.LP
A companion program
.RI ( unzip (1L))
unpacks
.I zip
archives.
The
.I zip
and
.IR unzip (1L)
programs can work with archives produced by most zip utilities
(depending on the features used) and most unzip utilities can
work with archives produced by \fIzip\fP (though some utilities
may still not support streamed archives yet).  \fIzip\fP
supports most features of the zip standard up to PKZIP version
4.6.
.I zip
version 3.1 supports the Zip64 extensions of PKZIP 4.5 which
allow archives as well as files to exceed the previous 2 GB limit
(4 GB in some cases).  \fIzip\fP also now supports \fBbzip2\fP
compression if the \fBbzip2\fP library is included when \fIzip\fP
is compiled.  Note that PKUNZIP 1.10 cannot extract files produced
by PKZIP 2.04 or \fIzip\ 2.0\fP (or later).  You must use PKUNZIP
2.04g or \fIunzip\ 5.0p1\fP (or later versions) to extract them.
.PP
See the \fBEXAMPLES\fP section at the bottom of this page
for examples of some typical uses of \fIzip\fP.
.PP
\fBLarge\ Archives\ and\ Zip64.\fP
.I zip
automatically uses the Zip64 extensions when files larger than 4 GB are
added to an archive, an archive containing Zip64 entries is updated
(if the resulting archive still needs Zip64),
the size of the archive will exceed 4 GB, or when the
number of entries in the archive will exceed about 64K.
Zip64 is also used for archives streamed from standard input as the size
of such archives are not known in advance, but the option \fB\-fz\-\fP can
be used to force \fIzip\fP to create PKZIP 2 compatible archives (as long
as Zip64 extensions are not needed).  You must use a PKZIP 4.5
compatible unzip, such as \fIunzip\ 6.0\fP or later, to extract files
using the Zip64 extensions.
.PP
\fBStreamed\ archives.\fP
Streamed archives, entries encrypted with standard encryption,
or split archives created with the pause option may not be compatible with
some utilities as data descriptors are used and they do not yet support
them, though the latest PKWare published zip standard now includes the data
descriptor format \fIzip\fP uses.  Other utilities do support these archives,
however.  If you run into problems, \fIzip\fP provides a way to strip data
descriptors from an existing archive using the \fB\-\-copy\fP option (if
standard encryption was not used).

.PP
\fBMac OS X.\fP  Though previous Mac versions had their own \fIzip\fP port,
\fIzip\fP supports Mac OS X as part of the Unix port and most Unix features
apply.  References to "MacOS" below generally refer to MacOS versions older
than OS X.  Mac OS X features supported by the Unix port, such
as resource forks, are identified as UNIX APPLE features.

.PP
For a brief help on \fIzip\fP and \fIunzip\fP,
run each without specifying any parameters on the command line.

.SH "USE"
.PP
The program is useful for packaging a set of files for distribution;
for archiving files;
and for saving disk space by temporarily
compressing unused files or directories.
.LP
The
.I zip
program puts one or more compressed files into a single
.I zip
archive,
along with information about the files
(name, path, date, time of last modification, protection,
and check information to verify file integrity).
An entire directory structure can be packed into a
.I zip
archive with a single command.
Compression ratios of 2:1 to 3:1 are common for text files.
.I zip
has one compression method (deflation) and can also store files without
compression.  (If \fBbzip2\fP support is added, \fIzip\fP can also
compress using \fBbzip2\fP compression, but such entries require a
reasonably modern unzip to decompress.  When \fBbzip2\fP compression
is selected, it replaces deflation as the default method.)
.I zip
automatically chooses the better of the two (deflation or store, or, if
\fBbzip2\fP is selected, \fBbzip2\fP or store) for each file to be
compressed.
.LP
\fBCommand\ format.\fP  The basic command format is
.IP
\fBzip\fR options archive inpath inpath ...
.LP
where \fBarchive\fR is a new or existing \fIzip\fR archive
and \fBinpath\fR is a directory or file path optionally including wildcards.
When given the name of an existing
.I zip
archive,
.I zip
will replace identically named entries in the
.I zip
archive (matching the relative names as stored in
the archive) or add entries for new names.
For example,
if
.I foo.zip
exists and contains
.I foo/file1
and
.IR foo/file2 ,
and the directory
.I foo
contains the files
.I foo/file1
and
.IR foo/file3 ,
then:
.IP
\fCzip -r foo.zip foo\fP
.LP
or more concisely
.IP
\fCzip -r foo foo\fP
.LP
will replace
.I foo/file1
in
.I foo.zip
and add
.I foo/file3
to
.IR foo.zip .
After this,
.I foo.zip
contains
.IR foo/file1 ,
.IR foo/file2 ,
and
.IR foo/file3 ,
with
.I foo/file2
unchanged from before.
.LP
So if before the zip command is executed \fIfoo.zip\fP has:
.IP
\fC foo/file1 foo/file2
.LP
and directory foo has:
.IP
\fC file1 file3\fP
.LP
then \fIfoo.zip\fP will have:
.IP
\fC foo/file1 foo/file2 foo/file3\fP
.LP
where \fIfoo/file1\fP is replaced and
\fIfoo/file3\fP is new.
.LP
\fB\-@\ file\ lists.\fP  If a file list is specified as
\fB\-@\fP
[Not on MacOS],
.I zip
takes the list of input files from standard input instead of from
the command line.  For example,
.IP
\fCzip -@ foo\fP
.LP
will store the files listed one per line on stdin in \fIfoo.zip\fP.
.LP
Under Unix,
this option can be used to powerful effect in conjunction with the
\fIfind\fP\ (1)
command.
For example,
to archive all the C source files in the current directory and
its subdirectories:
.IP
\fCfind . -name "*.[ch]" -print | zip source -@\fP
.LP
(note that the pattern must be quoted to keep the shell from expanding it).
.LP
\fBStreaming\ input\ and\ output.\fP
.I zip
will also accept a single dash ("-") as the zip file name, in which case it
will write the zip file to standard output, allowing the output to be piped
to another program. For example:
.IP
\fCzip -r - . | dd of=/dev/nrst0 obs=16k\fP
.LP
would write the zip output directly to a tape with the specified block size
for the purpose of backing up the current directory.
.LP
.I zip
also accepts a single dash ("-") as the name of a file to be compressed, in
which case it will read the file from standard input, allowing zip to take
input from another program. For example:
.IP
\fCtar cf - . | zip backup -\fP
.LP
would compress the output of the tar command for the purpose of backing up
the current directory. This generally produces better compression than
the previous example using the -r option because
.I zip
can take advantage of redundancy between files. The backup can be restored
using the command
.IP
\fCunzip -p backup | tar xf -\fP
.LP
When no zip file name is given and stdout is not a terminal,
.I zip
acts as a filter, compressing standard input to standard output.
For example,
.IP
\fCtar cf - . | zip | dd of=/dev/nrst0 obs=16k\fP
.LP
is equivalent to
.IP
\fCtar cf - . | zip - - | dd of=/dev/nrst0 obs=16k\fP
.LP
.I zip
archives created in this manner can be extracted with the program
.I funzip
which is provided in the
.I unzip
package, or by
.I gunzip
which is provided in the
.I gzip
package (but some
.I gunzip
may not support this if
.I zip
used the Zip64 extensions). For example:
.IP
\fPdd if=/dev/nrst0  ibs=16k | funzip | tar xvf -\fC
.LP
The stream can also be saved to a file and
.I unzip
used.
.LP
If Zip64 support for large files and archives is enabled and
\fIzip\fR is used as a filter, \fIzip\fR creates a Zip64 archive
that requires a PKZIP 4.5 or later compatible unzip to read it.  This is
to avoid amgibuities in the zip file structure as defined in the current
zip standard (PKWARE AppNote) where the decision to use Zip64 needs to
be made before data is written for the entry, but for a stream the size
of the data is not known at that point.  If the data is known to be smaller
than 4 GB, the option \fB\-fz\-\fP can be used to prevent use of Zip64,
but \fIzip\fP will exit with an error if Zip64 was in fact needed.
\fIzip\ 3\fR and \fIunzip\ 6\fR and later can read archives with Zip64
entries.  Also, \fIzip\fP removes the Zip64 extensions if not needed
when archive entries are copied (see the \fB\-U\fP (\fB\-\-copy\fP)
option).
.LP
When directing the output to another file, note that all options should be
before the redirection including \fB-x\fP.  For example:
.IP
\fPzip archive "*.h" "*.c" -x donotinclude.h orthis.h > tofile\fC
.LP

\fBZip\ files.\fP  When changing an existing
.I zip
archive,
.I zip
will write a temporary file with the new contents,
and only replace the old one when the process of creating the new version
has been completed without error.  (An exception is the \fB\-g\fP
(\fB\-\-grow\fP) option, which appends to the original archive.)
.LP
If the name of the
.I zip
archive does not contain an extension, the extension
\fB.zip\fP
is added. If the name already contains an extension other than
\fB.zip\fP,
the existing extension is kept unchanged.  However, split archives
(archives split over multiple files) require the \fB.zip\fP extension
on the last split.

.PP
\fBScanning\ and\ reading\ files.\fP
When \fIzip\fP starts, it scans for files to process (if needed).  If
this scan takes longer than about 5 seconds, \fIzip\fP will display
a "Scanning files" message and start displaying progress dots every 2 seconds
or every so many entries processed, whichever takes longer.  If there is more
than 2 seconds between dots it could indicate that finding each file is taking
time and could mean a slow network connection for example.
(Actually the initial file scan is
a two-step process where the directory scan is followed by a sort and these
two steps are separated with a space in the dots.  If updating an existing
archive, a space also appears between the existing file scan and the new
file scan.)  The scanning files dots are not controlled by the \fB\-ds\fP
dot size option, but the dots are turned off by the \fB\-q\fP quiet option.  The
\fB\-sf\fP show files option can be used to scan for files and get the list of
files scanned without actually processing them.
.LP
If \fIzip\fR is not able to read a file, it
issues a warning but
continues.  See the \fB\-MM\fP option below for more on how \fIzip\fP handles
patterns that are not matched and files that are not readable.
If some files were skipped, a
warning is issued at the end of the zip operation noting how many files
were read and how many skipped.
.PP
\fBCommand\ modes.\fP  \fIzip\fP now supports two distinct types of command
modes, \fBexternal\fP and \fBinternal\fP.  The \fBexternal\fP modes
(add, update, and freshen) read files from the file system (as well as from an
existing archive) while the \fBinternal\fP modes (delete and copy) operate
exclusively on entries in an existing archive.
.LP
.TP
.BI add\ \ \ \ \ \ 
Update existing entries and add new files.  If the archive does not exist
create it.  This is the default mode.
.TP
.BI update\ \fP(\fB\-u\fP)
Update existing entries if newer on the file system and add new files.  If
the archive does not exist, issue a warning and then create a new archive.
.TP
.BI freshen\ \fP(\fB\-f\fP)
Update existing entries of an archive if newer on the file system.
Does not add new files to the archive.
.TP
.BI delete\ \fP(\fB\-d\fP)
Select entries in an existing archive and delete them.
.TP
.BI copy\ \fP(\fB\-U\fP)
Select entries in an existing archive and copy them to a new archive.
This new mode is similar to \fBupdate\fP but command line patterns
select entries in the existing archive rather than files from
the file system and it uses the \fB\-\-out\fP option to write the
resulting archive to a new file rather than update the existing
archive, leaving the original archive unchanged.
.LP
The new File Sync option (\fB\-FS\fP) is also considered a new mode,
though it is similar to \fBupdate\fP.  This mode synchronizes the
archive with the files on the OS, only replacing files in the
archive if the file time or size of the OS file is different, adding
new files, and deleting entries from the archive where there is
no matching file.  As this mode can delete entries from the archive,
consider making a backup copy of the archive.

Also see \fB\-DF\fP for creating difference archives.

See each option description below for details and the \fBEXAMPLES\fP section
below for examples.
.PP
\fBSplit\ archives.\fP  \fIzip\fP version 3.0 and later can create split
archives.  A
\fBsplit archive\fP is a standard zip archive split over multiple
files.  (Note that split archives are not just archives split in to
pieces, as the offsets of entries are now based on the start of each
split.  Concatenating the pieces together will invalidate these offsets,
but \fIunzip\fP can usually deal with it.  \fIzip\fP will usually refuse
to process such a spliced archive unless the \fB\-FF\fP fix option is
used to fix the offsets.)
.LP
One use of split archives is storing a large archive on multiple
removable media.
For a split archive with 20 split files the files are typically named (replace
ARCHIVE with the name of your archive) ARCHIVE.z01, ARCHIVE.z02, ..., ARCHIVE.z19,
ARCHIVE.zip.  Note that the last file is the \fB.zip\fP file.  In contrast,
\fBspanned archives\fP are the original multi-disk archive generally requiring
floppy disks and using volume labels to store disk numbers.  \fIzip\fP supports
split archives but not spanned archives, though a procedure exists for converting
split archives of the right size to spanned archives.  The reverse is also true,
where each file of a spanned archive can be copied in order to files with the
above names to create a split archive.
.LP
Use \fB\-s\fP to set the split size and create a split archive.  The size is
given as a number followed optionally by one of k (kB), m (MB), g (GB), or t (TB)
(the default is m).  The \fB\-sp\fP option can be used to pause \fIzip\fP between
splits to allow changing removable media, for example, but read the descriptions
and warnings for both \fB\-s\fP and \fB\-sp\fP below.
.LP
Though \fIzip\fP does not update split archives, \fIzip\fP provides the new
option \fB\-O\fP (\fB\-\-output\-file\fP or \fB\-\-out\fP) to allow split archives
to be updated and saved in a new archive.  For example,
.IP
\fCzip inarchive.zip foo.c bar.c \-\-out outarchive.zip\fP
.LP
reads archive \fBinarchive.zip\fP, even if split, adds the files \fBfoo.c\fP and
\fBbar.c\fP, and writes the resulting archive to \fBoutarchive.zip\fP.  If
\fBinarchive.zip\fP is split then \fBoutarchive.zip\fP defaults to the same
split size.  Be aware that if \fBoutarchive.zip\fP and any split files that are
created with it already exist, these are always overwritten as needed without
warning.  This may be changed in the future.
.PP
\fBUnicode.\fP  Though the zip standard requires storing paths in an
archive using a specific character set, in practice zips have stored
paths in archives in whatever the local character set is.  This creates
problems when an archive is created or updated on a system using one
character set and then extracted on another system using a different
character set.  When compiled with Unicode support enabled on platforms
that support wide characters, \fIzip\fP now stores, in addition to the
standard local path for backward compatibility, the UTF-8 translation of
the path.  This provides a common universal character set for storing
paths that allows these paths to be fully extracted on other systems that
support Unicode and to match as closely as possible on systems that
don't.

On Win32 systems where paths are internally stored as Unicode but represented in
the local character set, it's possible that some paths will be skipped during a
local character set directory scan.  \fIzip\fP with Unicode support now can read
and store these paths.  Note that Win 9x systems and FAT file systems don't fully
support Unicode.

Be aware that console windows on Win32 and Unix, for example, sometimes don't
accurately show all characters due to how each operating system switches in
character sets for display.  However, directory navigation tools should show the
correct paths if the needed fonts are loaded.
.PP
\fBCommand line format.\fP  This version of
.I zip
has updated command line processing and support for long options.
.PP
Short options take the form
.IP
\fC-s[-][s[-]...][value][=value][\ value]\fP
.LP
where s is a one or two character short option.  A short option
that takes a value is last in an argument and anything after it is
taken as the value.  If the option can be negated and "-" immediately
follows the option, the option is negated.
Short options can also be given as separate arguments
.IP
\fC-s[-][value][=value][\ value]\ -s[-][value][=value][\ value]\ ...\fP
.LP
Short options in general take values either as part of the same
argument or as the following argument.  An optional = is also supported.
So
.IP
\fC-ttmmddyyyy\fP
.LP
and
.IP
\fC-tt=mmddyyyy\fP
.LP
and
.IP
\fC-tt mmddyyyy\fP
.LP
all work.  The \fB\-x\fP and \fB\-i\fP options accept lists of values
and use a slightly different format described below.  See the
\fB\-x\fP and \fB\-i\fP options.
.PP
Long options take the form
.IP
\fC--longoption[-][=value][ value]\fP
.LP
where the option starts with --, has a multicharacter name, can
include a trailing dash to negate the option (if the option
supports it), and can have a value (option argument) specified by
preceeding it with = (no spaces).  Values can also follow the
argument.  So
.IP
\fC--before-date=mmddyyyy\fP
.LP
and
.IP
\fC--before-date mmddyyyy\fP
.LP
both work.

Long option names can be shortened to the shortest unique
abbreviation.  See the option descriptions below for which
support long options.  To avoid confusion, avoid abbreviating
a negatable option with an embedded dash ("-") at the dash
if you plan to negate it (the parser would consider
a trailing dash, such as for the option \fB\-\-some\-option\fP using
\fB\-\-some\-\fP as the option, as part of the name rather
than a negating dash).  This may be changed to force the last
dash in \fB\-\-some\-\fP to be negating in the future.
.SH "OPTIONS"
.TP
.PD 0
.BI \-a
.TP
.PD
.B \-\-ascii
[Systems using EBCDIC] Translate file to ASCII format.

.TP
.PD 0
.BI \-aa
.TP
.PD
.B \-\-all-ascii
Handle all files as ASCII text files.  This is mainly to
support EBCDIC/ASCII conversions.

.TP
.PD 0
.B \-A
.TP
.PD
.B \-\-adjust-sfx
Adjust self-extracting executable archive.
A self-extracting executable archive is created by prepending
the SFX stub to an existing archive. The
.B \-A
option tells
.I zip
to adjust the entry offsets stored
in the archive to take into account this "preamble" data.
.LP
Note: self-extracting archives for the Amiga are a special case.
At present, only the Amiga port of \fIzip\fP is capable of adjusting
or updating these without corrupting them. -J can be used to remove
the SFX stub if other updates need to be made.

.TP
.PD 0
.B \-AC
.TP
.PD
.B \-\-archive-clear
[WIN32]  Once archive is created (and tested if \fB\-T\fP is used,
which is recommended), clear the archive bits of files processed.  WARNING:
Once the bits are cleared they are cleared.  You may want to use the
\fB\-sf\fP show files option to store the list of files processed in case
the archive operation must be repeated.  Also consider using
the \fB\-MM\fP must match option.  Be sure to check out \fB\-DF\fP as a
possibly better way to do incremental backups.

.TP
.PD 0
.B \-AS
.TP
.PD
.B \-\-archive-set
[WIN32]  Only include files that have the archive bit set.  Directories
are not stored when \fB\-AS\fP is used, though by default the paths
of entries, including directories, are stored as usual and can be used
by most unzips to recreate directories.

The archive bit is set by the operating system when a file is modified
and, if used with \fB\-AC\fP, \fB\-AS\fP can provide an
incremental backup capability.  However, other applications can
modify the archive bit and it may not be a reliable indicator of
which files have changed since the last archive operation.  Alternative
ways to create incremental backups are using \fB\-t\fP to use file dates,
though this won't catch old files copied to directories being archived,
and \fB\-DF\fP to create a differential archive.

.TP
.PD 0
.B \-as
.TP
.PD
.B \-\-sequester
[UNIX APPLE]  Save Apple Double resource files in a sequestered directory,
instead of in the directory where the data fork file is.

.TP
.PD 0
.B \-B
.TP
.PD
.B \-\-binary
[VM/CMS and MVS] force file to be read binary (default is text).

.TP
.B \-B\fRn
[TANDEM] set Edit/Enscribe formatting options with n defined as
.RS
bit  0: Don't add delimiter (Edit/Enscribe)
.RE
.RS
bit  1: Use LF rather than CR/LF as delimiter (Edit/Enscribe)
.RE
.RS
bit  2: Space fill record to maximum record length (Enscribe)
.RE
.RS
bit  3: Trim trailing space (Enscribe)
.RE
.RS
bit  8: Force 30K (Expand) large read for unstructured files
.RE

.TP
.PD 0
.BI \-b\ \fRpath
.TP
.PD
.B \-\-temp-path\ \fRpath
Use the specified
.I path
for the temporary
.I zip
archive. For example:
.RS
.IP
\fCzip -b /tmp stuff *\fP
.RE
.IP
will put the temporary
.I zip
archive in the directory
.IR /tmp ,
copying over
.I stuff.zip
to the current directory when done. This option is useful when
updating an existing archive and the file system containing this
old archive does not have enough space to hold both old and new archives
at the same time.  It may also be useful when streaming in some
cases to avoid the need for data descriptors.  Note that using
this option may require \fIzip\fP take additional time to copy
the archive file when done to the destination file system.

.TP
.PD 0
.B \-c
.TP
.PD
.B \-\-entry-comments
Add one-line comments for each file.
File operations (adding, updating) are done first,
and the user is then prompted for a one-line comment for each file.
Enter the comment followed by return, or just return for no comment.

.TP
.PD 0
.B \-C
.TP
.PD
.B \-\-preserve-case
[VMS]  Preserve case all on VMS.  Negating this option
(\fB\-C-\fP) downcases.

.TP
.PD 0
.B \-C2
.TP
.PD
.BI \-\-preserve-case-2
[VMS]  Preserve case ODS2 on VMS.  Negating this option
(\fB\-C2\-\fP) downcases.  Default:  \-C2\- (downcase ODS2 names).

.TP
.PD 0
.B \-C5
.TP
.PD
.B \-\-preserve-case-5
[VMS]  Preserve case ODS5 on VMS.  Negating this option
(\fB\-C5\-\fP) downcases.  Default:  \-C5 (preserve case of ODS5 names).

.TP
.PD 0
.B \-d
.TP
.PD
.B \-\-delete

Source/emergeLib/zip.c
/*
   Copyright (c) 1990-2010 Info-ZIP.  All rights reserved.

   See the accompanying file LICENSE, version 2009-Jan-02 or later
   (the contents of which are also included in zip.h) for terms of use.
   If, for some reason, all these files are missing, the Info-ZIP license
   also may be found at:  ftp://ftp.info-zip.org/pub/infozip/license.html
   */
/*
   A very simplistic example of how to load the zip dll and make a call into it.
   Note that none of the command line options are implemented in this example.

*/

#ifndef WIN32
#  define WIN32
#endif
#define API

/* Tell Microsoft Visual C++ 2005 to leave us alone and
 * let us use standard C functions the way we're supposed to.
 */
#if defined(_MSC_VER) && (_MSC_VER >= 1400)
#  ifndef _CRT_SECURE_NO_DEPRECATE
#    define _CRT_SECURE_NO_DEPRECATE
#  endif
#  ifndef _CRT_NONSTDC_NO_DEPRECATE
#    define _CRT_NONSTDC_NO_DEPRECATE
#  endif
#endif

/*#include <stdlib.h>
#include <stdio.h>
#include <commdlg.h>
#include <dlgs.h>*/

#ifndef EXPENTRY
#define EXPENTRY WINAPI
#endif

#include "../zip/windll/structs.h"

/* Defines */
#ifndef MSWIN
#define MSWIN
#endif

typedef int (WINAPI * _DLL_ZIP)(ZCL, LPZPOPT);
typedef int (WINAPI * _ZIP_USER_FUNCTIONS)(LPZIPUSERFUNCTIONS);

/* Global variables */

extern LPZIPUSERFUNCTIONS lpZipUserFunctions;

extern HINSTANCE hZipDll;

extern int hFile;                 /* file handle             */

/* Global functions */

extern int WINAPI ZpArchive(ZCL C, LPZPOPT Opts);
extern int WINAPI ZpInit(LPZIPUSERFUNCTIONS lpZipUserFunc);

#include <assert.h>    /* required for all Windows applications */
#include <windowsx.h>
#include <windows.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <time.h>
#include <string.h>
#ifdef __BORLANDC__
#include <dir.h>
#else
#include <direct.h>
#endif
#include "zip.h"
#include "../zip/revision.h"
/* printf get redirected by a macro define in api.h !!! */
#ifdef printf
# undef printf
#endif

#ifdef WIN32
#include <commctrl.h>
#include <winver.h>
#else
#include <ver.h>
#endif

#ifdef __GNUC__
#define UNUSED __attribute__((unused))
#else
#define UNUSED
#endif

#ifdef WIN32
#define ZIP_DLL_NAME "ZIP32Z64.DLL\0"
#else
#define ZIP_DLL_NAME "ZIP16.DLL\0"
#endif

#define DLL_WARNING "Cannot find %s."\
  " The Dll must be in the application directory, the path, "\
"the Windows directory or the Windows System directory."
#define DLL_VERSION_WARNING "%s has the wrong version number."\
  " Insure that you have the correct dll's installed, and that "\
"an older dll is not in your path or Windows System directory."

int hFile;              /* file handle */

ZCL ZpZCL;
LPZIPUSERFUNCTIONS lpZipUserFunctions;
HANDLE hZUF = (HANDLE)NULL;
HINSTANCE hUnzipDll;
HANDLE hFileList;
ZPOPT ZpOpt;
#ifdef WIN32
DWORD dwPlatformId = 0xFFFFFFFF;
#endif
HINSTANCE hZipDll;

/* Forward References */
_DLL_ZIP ZipArchive;
_ZIP_USER_FUNCTIONS ZipInit;

void FreeUpMemory(void);
int WINAPI DummyPassword(LPSTR, int, LPCSTR, LPCSTR);
int WINAPI DummyPrint(char far *, unsigned long);
int WINAPI WINAPI DummyComment(char far *);

#ifdef WIN32
BOOL IsNT(VOID);
#endif

/****************************************************************************

FUNCTION: Main(int argc, char **argv)

 ****************************************************************************/
#ifdef __BORLANDC__
#  ifdef WIN32
#pragma argsused
#  endif
#endif
int MakeZip(char *filename, char *root, char *path)
{
  /*char szFileList[1][MAX_PATH];
    char testPath[MAX_PATH];*/
  LPSTR szFileList;
  char **index, *sz;
  int retcode, i, cc;
  DWORD dwVerInfoSize;
  DWORD dwVerHnd;
  char szFullPath[PATH_MAX];
#ifdef WIN32
  char *ptr;
#else
  HFILE hfile;
  OFSTRUCT ofs;
#endif
  HANDLE  hMem;         /* handle to mem alloc'ed */

  hZUF = GlobalAlloc( GPTR, (DWORD)sizeof(ZIPUSERFUNCTIONS));
  if (!hZUF)
    {
      return 0;
    }
  lpZipUserFunctions = (LPZIPUSERFUNCTIONS)GlobalLock(hZUF);

  if (!lpZipUserFunctions)
    {
      GlobalFree(hZUF);
      return 0;
    }

  lpZipUserFunctions->print = DummyPrint;
  lpZipUserFunctions->password = DummyPassword;
  lpZipUserFunctions->comment = DummyComment;

  if (!ZpInit(lpZipUserFunctions))
    {
      printf("Application functions not set up properly. Terminating...");
      FreeUpMemory();
      return 0;
    }

  /* Here is where the action starts */
  memset(&ZpOpt, 0, sizeof(ZpOpt));
  ZpOpt.ExcludeBeforeDate = NULL;    /* set to valid Zip date, or NULL */
  ZpOpt.IncludeBeforeDate = NULL;    /* set to valid Zip date, or NULL */
  ZpOpt.szRootDir = root;            /* set to root dir (will cd to), or NULL */
  ZpOpt.szTempDir = NULL;            /* set to dir for temp files, or NULL */
  ZpOpt.fUnicode = 4;                /* Unicode flag */
  ZpOpt.fEncrypt = FALSE;            /* Encrytion flag */
  ZpOpt.fSystem = FALSE;             /* true to include system/hidden files */
  ZpOpt.fVolume = FALSE;             /* true if storing volume label */
  ZpOpt.fExtra = FALSE;              /* true if including extra attributes */
  ZpOpt.fNoDirEntries = FALSE;       /* true if ignoring directory entries */
  ZpOpt.fVerbose = FALSE;            /* true if full messages wanted */
  ZpOpt.fQuiet = TRUE;               /* true if minimum messages wanted */
  ZpOpt.fCRLF_LF = FALSE;            /* true if translate CR/LF to LF */
  ZpOpt.fLF_CRLF = FALSE;            /* true if translate LF to CR/LF */
  ZpOpt.fJunkDir = FALSE;            /* true if junking directory names */
  ZpOpt.fGrow = FALSE;               /* true if allow appending to zip file */
  ZpOpt.fForce = FALSE;              /* true if making entries using DOS names */
  ZpOpt.fMove = FALSE;               /* true if deleting files added or updated */
  ZpOpt.fDeleteEntries = FALSE;      /* true if deleting files from archive */
  ZpOpt.fUpdate = FALSE;             /* true if updating zip file--overwrite only
                                        if newer */
  ZpOpt.fFreshen = FALSE;            /* true if freshening zip file--overwrite only */
  ZpOpt.fJunkSFX = FALSE;            /* true if junking sfx prefix*/
  ZpOpt.fLatestTime = FALSE;         /* true if setting zip file time to time of
                                        latest file in archive */
  ZpOpt.fComment = FALSE;            /* true if putting comment in zip file */
  ZpOpt.fOffsets = FALSE;            /* true if updating archive offsets for sfx
                                        files */
  ZpOpt.fPrivilege = 0;
  ZpOpt.fEncryption = 0;
  ZpOpt.szSplitSize = NULL;

  ZpOpt.szIncludeList = NULL;
  ZpOpt.IncludeListCount = 0;
  ZpOpt.IncludeList = NULL;
  ZpOpt.szExcludeList = NULL;
  ZpOpt.ExcludeListCount = 0;
  ZpOpt.ExcludeList = NULL;

  ZpOpt.fRecurse = 1;           /* subdir recursing mode: 1 = "-r", 2 = "-R" */
  ZpOpt.fRepair = 0;            /* archive repair mode: 1 = "-F", 2 = "-FF" */
  ZpOpt.fLevel = '6';           /* Default deflate compression level */
  ZpOpt.szCompMethod = NULL;
  for (i = 0; i < 8; i++) {
    ZpOpt.fluff[i] = 0;
  }
  getcwd(szFullPath, PATH_MAX); /* Set directory to current directory */

  ZpZCL.argc = 1;
  /*ZpZCL.argc = argc - 2;*/        /* number of files to archive - adjust for the
                                       actual number of file names to be added */
  ZpZCL.lpszZipFN = filename;    /* archive to be created/updated */

  /* Copy over the appropriate portions of argv, basically stripping out argv[0]
     (name of the executable) and argv[1] (name of the archive file)
     */
  hFileList = GlobalAlloc( GPTR, 0x10000L);
  if ( hFileList )
    {
      szFileList = (char far *)GlobalLock(hFileList);
    }
  index = (char **)szFileList;
  cc = (sizeof(char *) * ZpZCL.argc);
  sz = szFileList + cc;

  cc = lstrlen(path);
  lstrcpy(sz, path);
  index[0] = sz;
  sz += (cc + 1);

  ZpZCL.FNV = (char **)szFileList;  /* list of files to archive */

  /*lstrcpy(szFileList[0], path);
    ZpZCL.FNV = (char **)szFileList;*/

  /* Go zip 'em up */
  retcode = ZpArchive(ZpZCL, &ZpOpt);
  if (retcode != 0)
    printf("Error in archiving\n");


  GlobalUnlock(hFileList);
  GlobalFree(hFileList);
  FreeUpMemory();
  FreeLibrary(hZipDll);
  return 1;
}

void FreeUpMemory(void)
{
  if (hZUF)
    {
      GlobalUnlock(hZUF);
      GlobalFree(hZUF);
    }
}

#ifdef WIN32
/* This simply determines if we are running on NT */
BOOL IsNT(VOID)
{
  if(dwPlatformId != 0xFFFFFFFF)
    return dwPlatformId;
  else
    /* note: GetVersionEx() doesn't exist on WinNT 3.1 */
    {
      if(GetVersion() < 0x80000000)
        {
          dwPlatformId = TRUE;
        }
      else
        {
          dwPlatformId = FALSE;
        }
    }
  return dwPlatformId;
}
#endif

/* Password entry routine - see password.c in the wiz directory for how
   this is actually implemented in Wiz. If you have an encrypted file,
   this will probably give you great pain. Note that none of the
   parameters are being used here, and this will give you warnings.
   */
int WINAPI DummyPassword(LPSTR p UNUSED, int n UNUSED, LPCSTR m UNUSED, LPCSTR name UNUSED)
{
  return 1;
}

/* Dummy "print" routine that simply outputs what is sent from the dll */
int WINAPI DummyPrint(char far *buf, unsigned long size)
{
  printf("%s", buf);
  return (unsigned int) size;
}


/* Dummy "comment" routine. See comment.c in the wiz directory for how
   this is actually implemented in Wiz. This will probably cause you
   great pain if you ever actually make a call into it.
   */
int WINAPI DummyComment(char far *szBuf)
{
  szBuf[0] = '\0';
  return TRUE;
}
